Backport cibuildwheel updates to v3.7.x

QuLogic opened the pull request:
## PR summary

We let this go for a while, so automated backports were failing.

## PR checklist
- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

ksunden commented:
Rerunning the cibw jobs because it failed in fairly early stages on windows, the linux build got a bit further, but was cancelled when the windows build failed... Going to see if it is reproducible

QuLogic commented:
This is some annoyance with having to build 3.12 wheels for our dependencies; I must not have backported that exactly right (as I didn't backport the sdist+wheel merger).

QuLogic commented:
It looks like PyPy started failing a while back due to https://github.com/scipy/oldest-supported-numpy/pull/78, but in `main` and `v3.8.x`, we switched to `numpy>=1.25` so we were okay there. I'm not sure if we want to backport that whole change for this last bugfix release, so I've just modified the PyPy wheels.
